modules/autoscaling-policy-metric-alarm-pair/main.tf

```diff
@@ -0,0 +1,59 @@
+variable "name" {
+  description = "Prefix of the AutoScaling Policy."
+}
+
+variable "asg_name" {
+  description = "Name of the AutoScaling Group that this policy would apply to."
+}
+
+variable "metric" {
+  description = "CPUUtilization or MemoryUtilization"
+}
+
+resource "aws_autoscaling_policy" "scale_up" {
+  name                   = "${var.name}-asp-up"
+  autoscaling_group_name = "${var.asg_name}"
+  adjustment_type        = "ChangeInCapacity"
+  cooldown               = "300"
+  scaling_adjustment     = "3"
+  policy_type            = "SimpleScaling"
+}
+
+resource "aws_cloudwatch_metric_alarm" "scale_up" {
+  alarm_name          = "${var.name}-scale-up-alarm"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "3"
+  metric_name         = "${var.metric}"
+  namespace           = "AWS/EC2"
+  period              = "60"
+  statistic           = "Average"
+  threshold           = "60"
+  dimensions {
+    AutoScalingGroupName = "${var.asg_name}"
+  }
+  alarm_actions = ["${aws_autoscaling_policy.scale_up.arn}"]
+}
+
+resource "aws_autoscaling_policy" "scale_down" {
+  name                   = "${var.name}-asp-down"
+  autoscaling_group_name = "${var.asg_name}"
+  adjustment_type        = "ChangeInCapacity"
+  cooldown               = "300"
+  scaling_adjustment     = "-2"
+  policy_type            = "SimpleScaling"
+}
+
+resource "aws_cloudwatch_metric_alarm" "scale_down" {
+  alarm_name          = "${var.name}-scale-down-alarm"
+  comparison_operator = "LessThanOrEqualToThreshold"
+  evaluation_periods  = "3"
+  metric_name         = "${var.metric}"
+  namespace           = "AWS/EC2"
+  period              = "60"
+  statistic           = "Average"
+  threshold           = "30"
```

Comment:
I believe it would be good for us to parametize these details we would likely want to tweak. Eg the `evaluation_periods`, `period`, `threshold`, `scaling_adjustment` , `cooldown`, should all be variables with reasonable defaults.

Also, the module creates scale up and down resources in the module, but in the example env, the module is used twice, once for scaling up and once for scaling down - We should pick one.. either the module covers both scenarios or is used twice to cover each scenario. I believe we would be slightly better to have the module focus on one scenario at a time, and used multiple times if that is needed.

Reply:
OK.

Reply:
It's very likely that you would want slightly different parameters for scaling up / down. Eg, maybe there is a shorter cooldown period, or usually there is a longer period for the metrics evaluation. At the same time, you are right, you would use both the up and down together, but I am just recommending the module be simpler and used/called more times - eg instead of one module with both scale up and down, there is one general module we call twice when we want to use it.